weldx/asdf/validators.py

```diff
@@ -95,6 +95,261 @@ def _unit_validator(
+
+
+    params
+    ------
+    list_expected:
+        Expected List to validate against
+
+    raises
+    ------
+    ValueError:
+        ValueError will be raised if an rule violation is found
+    """
+    validator = 0
+    for exp in list_expected:
+        if validator == 1:
+            if "(" not in str(exp):
+                raise ValueError(
+                    "Optional dimensions in the expected "
+                    "shape should only stand at the end/beginning."
+                )
+        elif validator == 2:
+            raise ValueError('After "..." should not be another dimension.')
+        elif "..." in str(exp):
+            if "..." != exp:
+                raise ValueError(
+                    f'"..." should not have additional properties:' f" {exp} was found."
+                )
+            validator = 2
+        elif "(" in str(exp):
+            validator = 1
+
+
+def _compare_lists(_list, list_expected):
+    """Compare two lists.
+
+    params
+    ------
+    _list:
+        List of Integer
+    list_expected:
+        List build by the rules in _custom_shape_validator
+    returns
+    -------
+    False:
+        when a dimension mismatch occurs
+    dict_values:
+        when no dimension mismatch occurs. Can be empty {}. Dictionary - keys: variable
+        names in the validation schemes. values: values of the validation schemes.
+    """
+    dict_values = dict()
+    # Compare List with expected List
+    stopper = 0
+    for i, exp in enumerate(list_expected):
+        # when "..." was found all the following dimensions are accepted
+        if stopper:
+            continue
+        # if "..." is found all the following dimensions are accepted
+        elif "..." in str(exp):
+            stopper = 1
+
+        elif "(" in str(exp):
+            if i < len(_list):
+                comparable = exp[exp.index("(") + 1 : exp.rindex(")")]
+                if comparable.isalnum() and not comparable.isnumeric():
+                    if comparable in dict_values:
+                        if _list[i] != dict_values[comparable]:
+                            return False
+                    else:
+                        dict_values[comparable] = _list[i]
+                elif not _compare(_list[i], comparable):
+                    return False
+
+        # all alphanumeric strings are OK - only numeric strings are not
+        # eg: "n", "n1", "n1234", "myasdfstring1337"
+        elif str(exp).isalnum() and not str(exp).isnumeric():
+            # if value is already saved in dict_values
+            if exp in dict_values:
+                # compare
+                if _list[i] != dict_values[exp]:
+                    # error found
+                    return False
+            else:
+                # add to dict_values
+                dict_values[exp] = _list[i]
+        else:
+            if i >= len(_list) or not _compare(_list[i], str(exp)):
+                return False
```

Comment:
Again a lot of unnecessary `elif`/`else`. Also, try to combine multiple nested ifs into one statement. Every `if` introduces a new indent, making the code hard to read and hides unnecessary `elif`/`else`. So I`d suggest simplifying the whole block.